source_code/main_mcu/src/LOGIC/logic_gui.c

```diff
@@ -303,24 +333,65 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
         utils_ascii_to_unicode(child_node->password, NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH);
         child_node->cust_char_password[NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH] = 0;
     }    
-    
+
+    /* Decrypt TOTPsecret */
+    logic_encryption_ctr_decrypt(child_node->TOTP.TOTPsecret_ct, child_node->TOTP.TOTPsecret_ctr, MEMBER_SIZE(TOTP_cred_node_t, TOTPsecret), FALSE);
+
     /* Temp vars for our main loop */
     BOOL first_function_run = TRUE;
-    int16_t text_anim_x_offset[2];
-    BOOL text_anim_going_right[2];
+
+    int16_t text_anim_x_offset[LOGIC_GUI_DISP_CRED_NUM_LINES_MAX];
+    BOOL text_anim_going_right[LOGIC_GUI_DISP_CRED_NUM_LINES_MAX];
     BOOL redraw_needed = TRUE;
     int16_t displayed_length;
-    BOOL scrolling_needed[2];
+    BOOL scrolling_needed[LOGIC_GUI_DISP_CRED_NUM_LINES_MAX];
     
     /* Reset temp vars */
     memset(text_anim_going_right, FALSE, sizeof(text_anim_going_right));
     memset(text_anim_x_offset, 0, sizeof(text_anim_x_offset));
     memset(scrolling_needed, FALSE, sizeof(scrolling_needed));
     
+    /* Generate TOTP */
+    cust_char_t TOTP_str[LOGIC_GUI_TOTP_STR_LEN];
+    memset(TOTP_str, 0, sizeof TOTP_str);
+    if (child_node->TOTP.TOTPsecretLen > 0)
```

Comment:
Which maxlen and why is that needed? br_hmac_key_init has no restriction on input key size. If the key size is greater that the block size it will be capped at digest size.. See main_mcu/src/BearSSL/src/mac/hmac.c line 70 ("key_len = br_digest_size(dig);"

Reply:
good point!